TLS 1.3: Add mbedtls_ssl_write_early_data() API

ronald-cron-arm opened the pull request:
## Description
Second PR of two addressing #6340 .

## PR checklist
- [x] **changelog** not required, will add one for TLS 1.3 early data feature as a whole
- [x] **backport** not required, new feature
- [x] **tests** provided

mpg commented:
Changed the base twice in order to force-refresh github's cached list of commits, so that commits from #8755 no longer show as new here now that #8755 has been merged.

mpg reviewed (commented):
Only a first pass, getting myself familiar with this.

Also, I think we should add a warning to the documentation of `write_early_data()` about the fact that EarlyData is not as well protected against replay than normal ApplicationData, so clients should not send data using that API if having that data replayed would be a problem (unless they know the server implements full replay protection). (Ideally quoting from the RFC.)

ronald-cron-arm commented on library/ssl_msg.c:
```diff
@@ -6058,6 +6058,81 @@ int mbedtls_ssl_write(mbedtls_ssl_context *ssl, const unsigned char *buf, size_t
+     * send at least one record of early data.
+     * Otherwise, resume the handshake and if the handshake sequence stops
+     * waiting for some message from the server, send early data if we can.
+     */
```
> I tend to reason in terms of handshake steps. So, I understand that if we haven't sent our ClientHello yet, clearly we want to send that first, and that's what calling `mbedtls_ssl_handshake_step()` seems to be doing in the first branch. But I don't really understand why we want to call that when status is `SENT` - to me that means we've sent the extension hence the ClientHello already? Or is it that we've written it to the buffer but it hasn't actually been sent out on the network yet (in which case the name would be confusing).

That's the difference between MBEDTLS_SSL_EARLY_DATA_STATUS_SENT and MBEDTLS_SSL_EARLY_DATA_STATUS_CAN_WRITE. If the middlebox compatibility is enabled, after sending the ClientHello and before switching to the early data key for outbound traffic we have to send in clear the dummy CCS and that's one more handshake step.

ronald-cron-arm replied:
> Regarding the second branch: so, despite its name, this function's goal is not just to sent early data, it's to complete the handshake if possible, and if not because we're waiting for the server, sent early data. Right?

I wouldn't say that to complete the handshake is a goal of this API, it is rather to keep progressing the handshake as long as we are in the window where early data can be sent (status CAN_WRITE and ACCEPTED). We do not want the client to keep sending early data and the handshake to be stalled. Thus each time the API is called in the "write window" we try to advance as much as possible the handshake, basically up to the point where we wait for some server message or we wait for some data to be actually sent.

ronald-cron-arm replied:
> Note: with the second branch I'm having the same problem mentally mapping `EARLY_DATA_STATUS_VALUES` to handshake states. I'm thinking `CAN_WRITE` right after we've sent the ClientHello and before we received EncryptedExtensions from the server, and `ACCEPTED` is after we've receive the extension from the server. Is that correct?

Yes that's correct but maybe `CAN_WRITE` right after we've sent the ClientHello as we may still have to send a dummy CCS (see first response above).

mpg replied:
> We do not want the client to keep sending early data and the handshake to be stalled.

Ah, good point. This makes the API more user-friendly in that users can just keep calling this function until it returns an error. Otherwise they'd need to be sure to call `handshake()` in-between calls to this function, which would be a bit awkward, and we're in a better position than users to know when that's useful. Thanks for clarifying that (and answering my other questions).

ronald-cron-arm commented on tests/suites/test_suite_ssl.function:
```diff
@@ -4101,3 +4101,203 @@ exit:
+                TEST_FAIL("Unexpected state.");
+        }
+
+        mbedtls_test_mock_socket_close(&(client_ep.socket));
```
Improved comment at the top of the loop.

ronald-cron-arm commented on programs/ssl/ssl_client2.c:
```diff
@@ -3035,6 +3035,55 @@ int main(int argc, char *argv[])
+
+                    /* For event-driven IO, wait for socket to become available */
+                    if (opt.event == 1 /* level triggered IO */) {
+#if defined(MBEDTLS_TIMING_C)
```
I'd prefer not changing that in this PR.

ronald-cron-arm commented:
@mpg I have addressed your comments, the Open CI is green, please have another look, thanks.

mpg commented:
Btw, out of scope for this PR, but I think the early data APIs are a bit complex (not saying they're bad, I think 0-RTT is just a fairly complex feature of 1.3 so this reflects in the APIs), and a knowledge base article explaining everything a user needs to know (in general, client-side and server-side) would be helpful if we can write one at some point.

ronald-cron-arm commented:
> Btw, out of scope for this PR, but I think the early data APIs are a bit complex (not saying they're bad, I think 0-RTT is just a fairly complex feature of 1.3 so this reflects in the APIs), and a knowledge base article explaining everything a user needs to know (in general, client-side and server-side) would be helpful if we can write one at some point.

We have already something at the end of `tls13-support.md`. The read part is outdated but the write part not that much it seems. I've planned to update/fix it when addressing the wrap-up issue.

mpg commented:
> We have already something at the end of `tls13-support.md`. The read part is outdated but the write part not that much it seems. I've planned to update/fix it when addressing the wrap-up issue.

Great! Though I'll note as a user I would probably not be looking under `docs/architecture`. Probably worth splitting the document into two parts: one for developers (same location), one for users (probably just under `docs`?).

mpg commented on tests/opt-testcases/tls13-misc.sh:
```diff
@@ -277,6 +277,31 @@ run_test    "TLS 1.3 m->G: EarlyData: basic check, good" \
+            -s "END OF EARLY DATA (5) was received." \
+            -s "early data accepted" \
+            -s "decrypted early data with length"
+
```
Notes:
- We appear to be only testing early data with GnuTLS, though it seems to be we should be able to test with ourselves too now. (Not expecting any issues, but as matter of principle, would be pretty embarrassing if we couldn't interop with ourselves.)
- In addition to the tests with `early_data=1` I'd like to see two reference tests (one with `early_data=0`, one without `early_data` option - so, library default), with the same list of assertions except now all negative (except "Reconnecting with saved session"), just to validate that we're really not sending it unless explicitly enabled at runtime (even when enabled at compile time. (Again, not expecting this to uncover issues, more as matter of principle.)

Edit: both tracked in #6362, so not a problem for this PR.

ronald-cron-arm replied:
Added a pointer to this comment in #6362 to be sure we cover it all.

mpg reviewed (commented):
Thanks for addressing my feedback! Looking pretty good, only minor points left.

Also, it seems to me we can now remove [the hack used for testing server before we had client support](https://github.com/Mbed-TLS/mbedtls/blob/development/tests/suites/test_suite_ssl.function#L36). Btw, now that we have a `tls13_write_early_data` test function, perhaps `tls13_early_data` should be renamed to `tls13_read_early_data`. Also, why doesn't that one have a `NO_INDICATION_SENT` scenario like the other two? (I'm OK with considering this whole paragraph out of scope as long as it's tracked somewhere.)

ronald-cron-arm commented on library/ssl_tls13_client.c:
```diff
@@ -2318,6 +2318,31 @@ static int ssl_tls13_write_end_of_early_data(mbedtls_ssl_context *ssl)
+            break;
+
+        default:
+            return MBEDTLS_ERR_SSL_INTERNAL_ERROR;
```
> Further, this made me think: how are users supposed to check the return value of this function? If it's not `BAD_INPUT_DATA` then success? Or if it's not one of the three `EARLY_DATA_STATUS` values then error?

If the returned value is negative then error otherwise you get the status. That seems okay to me.

mpg replied:
Ah no, you're right, `< 0` or `>= 0` is a good way to distinguish between failure and success.

ronald-cron-arm commented:
> Also, it seems to me we can now remove [the hack used for testing server before we had client support](https://github.com/Mbed-TLS/mbedtls/blob/development/tests/suites/test_suite_ssl.function#L36). 
I've removed it in this PR (not used anymore) but I will reintroduced it in max_early_data_size PRs for negative testing purposes.

>Btw, now that we have a `tls13_write_early_data` test function, perhaps `tls13_early_data` should be renamed to `tls13_read_early_data`. 
Done

>Also, why doesn't that one have a `NO_INDICATION_SENT` scenario like the other two? (I'm OK with considering this whole paragraph out of scope as long as it's tracked somewhere.)
Because not much happens I guess, but better with it thus I've added it.

ronald-cron-arm commented:
@mpg @waleed-elmelegy-arm this is ready for the next round of reviews, thanks.

mpg reviewed (approved):
LGTM, thanks for addressing my feedback and answering my questions!

ronald-cron-arm commented:
> LGTM, thanks for addressing my feedback and answering my questions!

Thanks for the review and the questions. @waleed-elmelegy-arm please have a look, thanks.

mpg commented:
> > Also, it seems to me we can now remove [the hack used for testing server before we had client support](https://github.com/Mbed-TLS/mbedtls/blob/development/tests/suites/test_suite_ssl.function#L36).
> 
> I've removed it in this PR (not used anymore) but I will reintroduced it in max_early_data_size PRs for negative testing purposes.

Perhaps we shouldn't remove it then. #8854 will no longer build (when merged with development) once this is merged.

ronald-cron-arm commented:
> > > Also, it seems to me we can now remove [the hack used for testing server before we had client support](https://github.com/Mbed-TLS/mbedtls/blob/development/tests/suites/test_suite_ssl.function#L36).
> > 
> > 
> > I've removed it in this PR (not used anymore) but I will reintroduced it in max_early_data_size PRs for negative testing purposes.
> 
> Perhaps we shouldn't remove it then. #8854 will no longer build (when merged with development) once this is merged.

I think if I keep it here the CI will fail on this PR because the function will not be used thus I guess we don't have really the choice.

mpg commented:
Ah, good point. We'll just need to rebase 8854 when this one gets merged, then.

waleed-elmelegy-arm commented on include/mbedtls/ssl.h:
```diff
@@ -734,6 +734,51 @@ typedef enum {
+ * The client has not sent the first ClientHello yet, it is unknown if the
+ * client will send an early data indication extension or not.
+ */
+    MBEDTLS_SSL_EARLY_DATA_STATUS_UNKNOWN,
```
I think this gives the idea that this is a state that track the actual early data not the "early data indication extension" and later on when reading the documentation or the code you realise that it's not, so maybe it's clearer to mention it in the name like with adding "_EXT_" or "_IND_" or in any other way that indicates it.

waleed-elmelegy-arm commented on include/mbedtls/ssl.h:
```diff
@@ -734,6 +734,51 @@ typedef enum {
+ * HelloRetryRequest) from the server yet. The transform to protect early data
+ * has been set and early data can be written now.
+ */
+    MBEDTLS_SSL_EARLY_DATA_STATUS_CAN_WRITE,
```
From the first glance it seems that "MBEDTLS_SSL_EARLY_DATA_STATUS_SENT" is the same as "MBEDTLS_SSL_EARLY_DATA_STATUS_CAN_WRITE" but it's not because of sending changeCipher compatibility so maybe it's better to mention it here in the comments why it's not the case since I was only able to know from the ticket discussion.

waleed-elmelegy-arm commented on include/mbedtls/ssl.h:
```diff
@@ -734,6 +734,51 @@ typedef enum {
+ */
+    MBEDTLS_SSL_EARLY_DATA_STATUS_NOT_SENT,
+    MBEDTLS_SSL_EARLY_DATA_STATUS_ACCEPTED,
+    MBEDTLS_SSL_EARLY_DATA_STATUS_REJECTED,
```
It seems to me that those three states are overlapping with the rest of the states for example "MBEDTLS_SSL_EARLY_DATA_STATUS_ACCEPTED" and "MBEDTLS_SSL_EARLY_DATA_STATUS_REJECTED" are both the same state as "MBEDTLS_SSL_EARLY_DATA_STATUS_SERVER_FINISHED_RECEIVED" but we still need to know the information of if the server accepted early data or not so maybe this can be split to maybe "Result" and "State" for example? or something close maybe.

mpg replied:
Actually should we just use different enums for the return value of this function and the internal state? I think that would be clearer.

waleed-elmelegy-arm reviewed (commented):
Thanks for the PR, just some comments for better clarity but nothing really wrong.

ronald-cron-arm commented:
@waleed-elmelegy-arm thanks for your comments. The PR is already quite big and I was thinking that I could address your comments in a follow-up PR. @mpg @waleed-elmelegy-arm would you be fine with that?

mpg commented:
Fine by me!
